core/utils/db.go

```diff
@@ -0,0 +1,107 @@
+package utils
+
+import (
+	"time"
+
+	"github.com/lib/pq"
+	"github.com/tevino/abool"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+)
+
+// PostgresEventListener listens to one type of postgres event as emitted by NOTIFY
+// TODO(sam): Currently each new listener opens a new connection. This is
+// suboptimal. We ought to move to a design similar to LogBroadcaster where we
+// hold only one open connection and listen to everything on that.
+type PostgresEventListener struct {
+	URI                  string
+	Event                string
+	PayloadFilter        string
+	MinReconnectInterval time.Duration
+	MaxReconnectDuration time.Duration
+
+	listener *pq.Listener
+	started  abool.AtomicBool
+	stopped  abool.AtomicBool
+	chEvents chan string
+	chStop   chan struct{}
+	chDone   chan struct{}
```

Comment:
I recommend receiving a `context.Context` instance in a constructor to simplify the termination logic!

Reply:
@spooktheducks

Reply:
Contexts shouldn't be used on long-running processes, only short-lived requests: https://godoc.org/context
> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.

The main reason is that you can't wait for the stopped service to shutdown fully after calling `cancel()` (which is something we endeavour to do in many of our `Stop()` methods).